File: cpp/202-HappyNumber/main.cpp
#include <iostream>
#include <unordered_set>

using namespace std;

class Solution 
{
private:
    int sumOfSquaredDigits(int n) const
    {
        int sum = 0;

        while(n)
        {
            auto digit = n % 10;
            sum += (digit * digit);

            n /= 10;
        }

        return sum;
    }

public:
    bool isHappy(int n) 
    {
        unordered_set<int> seen;

        while(n != 1)
        {
            auto insert_info = seen.insert(n);
            if(insert_info.second == false)
            {
                break;
            }

            n = sumOfSquaredDigits(n);
        }

        return n == 1;
    }
};

int main()
{
    Solution solution;
    std::cout << solution.isHappy(19) << "\n";
    std::cout << solution.isHappy(2) << "\n";
}

File: go/202-HappyNumber/main.go
package main

import "fmt"

func sumOfSquareDigits(n int) int {
	sum := 0

	for n != 0 {
		digit := n % 10
		sum += (digit * digit)

		n /= 10
	}

	return sum
}

func isHappy(n int) bool {
	seen := make(map[int]bool)

	for n != 1 {
		// If we've alread processed the number then there's
		// a cycle which we need to break out of
		if _, alreadyProcessed := seen[n]; alreadyProcessed {
			break
		}

		seen[n] = true
		n = sumOfSquareDigits(n)
	}

	return n == 1
}

func main() {
	fmt.Println(isHappy(19))
	fmt.Println(isHappy(2))
}
